Add CCZ4 functions for the gradient of the gradient of the lapse and the divergence of the lapse

macedo22 opened the pull request:
## Proposed changes

As part of implementing the first order CCZ4 system, this PR adds functionx for computing:
- the gradient of the gradient of the lapse
- the divergence of the lapse

The motivation for this is to compute eq. (21) and (22) in this [CCZ4 paper](https://arxiv.org/pdf/1707.09910.pdf), whose values will later be necessary for computing the evolved variables.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 committed:
Add CCZ4 grad_grad_lapse function

macedo22 committed:
Add CCZ4 divergence_lapse function

macedo22 commented:
@nilsdeppe Compute tags for these functions have been removed and it is now ready for review

nilsdeppe commented on src/Evolution/Systems/Ccz4/DerivLapse.hpp:
```diff
@@ -0,0 +1,43 @@
+ */
+template <size_t Dim, typename Frame, typename DataType>
+void grad_grad_lapse(
+    const gsl::not_null<tnsr::ij<DataType, Dim, Frame>*> grad_grad_lapse_,
```
`grad_grad_lapse_` -> `result` or `result_grad_grad_lapse`? We use the `_` suffix to denote private member variables

nilsdeppe commented on src/Evolution/Systems/Ccz4/DerivLapse.cpp:
```diff
@@ -48,6 +49,37 @@ tnsr::ij<DataType, Dim, Frame> grad_grad_lapse(
+    const tnsr::II<DataType, Dim, Frame>& inverse_conformal_metric,
+    const tnsr::ij<DataType, Dim, Frame>& grad_grad_lapse) noexcept {
+  destructive_resize_components(div_lapse, get_size(get(conformal_factor)));
+  for (auto& component : *div_lapse) {
```
isn't this for loop:
```cpp
get(div_lapse) = 0.0;
```
?

macedo22 replied:
Yes, I recycled this for loop from another function and didn't think about how it's not needed here

nilsdeppe commented on src/Evolution/Systems/Ccz4/DerivLapse.cpp:
```diff
@@ -48,6 +49,37 @@ tnsr::ij<DataType, Dim, Frame> grad_grad_lapse(
+          inverse_conformal_metric.get(i, j) * grad_grad_lapse.get(i, j);
+    }
+  }
+  div_lapse->get() *= square(conformal_factor.get());
```
`conformal_factor.get()` -> `get(conformal_factor)`

nilsdeppe reviewed (changes requested):
Overall looks good, couple minor things you can squash immediately

macedo22 commented:
@nilsdeppe Updated!

macedo22 commented:
@nilsdeppe Sorry to have dismissed your previous approval. Per your request on my other PR [here](https://github.com/sxs-collaboration/spectre/pull/3517#discussion_r710537178), I've just fixed the notation in this PR to be consistent by changing `\bar` -> `\tilde`, so it will need to be reapproved.

Edit: I've also fixed a couple other change requests from your review of that PR but here:
- Removed the `Tensor.hpp` include from `DerivLapse.hpp`
- Moved the include for `DerivLapse.hpp` to be above system includes
- Removed `TypeAliases.hpp` and `IndexType.hpp` includes from `DerivLapse.cpp`

nilsdeppe commented on src/Evolution/Systems/Ccz4/DerivLapse.hpp:
```diff
@@ -39,4 +39,34 @@ tnsr::ij<DataType, Dim, Frame> grad_grad_lapse(
+ *
+ * \details Computes the divergence as:
+ * \f{align}
+ *     \nabla^i \nabla_i \alpha &= \phi^2 \tilde{\gamma^{ij}}
```
You should have the `tilde` over the indices. That is `\tilde{\gamma}^{ij}` not `\tilde{\gamma^{ij}}`

nilsdeppe reviewed (commented):
You can squash immediately

macedo22 commented:
@nilsdeppe Updated and rebased!
